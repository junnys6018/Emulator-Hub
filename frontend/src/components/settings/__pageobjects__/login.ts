import { findElementBy } from '@/tests/selenium';
import { By } from 'selenium-webdriver';

export class LoginPage {
    private static addProfileLocator() {
        return By.css('button.add-profile');
    }

    private static saveProfileLocator() {
        return By.xpath("//button[text()='Save']");
    }

    private static profileIconLocator() {
        return By.css('button.profile-icon');
    }

    static clickAddProfile() {
        return findElementBy(LoginPage.addProfileLocator()).click();
    }

    static clickSaveButton() {
        return findElementBy(LoginPage.saveProfileLocator()).click();
    }

    static clickProfileIcon() {
        return findElementBy(LoginPage.profileIconLocator()).click();
    }
}
