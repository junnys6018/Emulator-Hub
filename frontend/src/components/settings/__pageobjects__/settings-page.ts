import { By } from 'selenium-webdriver';
import { findElementBy } from '@/tests/selenium';

export class SettingsPanel {
    private static get saveButtonLocator() {
        return By.xpath("//button[text()='Save']");
    }

    private static get resetAllButtonLocator() {
        return By.xpath("//button[text()='Reset All']");
    }

    static saveButton() {
        return findElementBy(SettingsPanel.saveButtonLocator);
    }

    static resetAllButton() {
        return findElementBy(SettingsPanel.resetAllButtonLocator);
    }
}

export class SettingsPage {
    private static sidePanelButtonLocator(panel: string) {
        return By.xpath(`//div[contains(@class, 'settings__button')]/button[text()='${panel}']`);
    }

    private static get titleLocator() {
        return By.css('h1');
    }

    static getTitle() {
        return findElementBy(SettingsPage.titleLocator).getText();
    }

    static clickGeneralSidePanelButton() {
        return findElementBy(SettingsPage.sidePanelButtonLocator('General')).click();
    }

    static clickNESSidePanelButton() {
        return findElementBy(SettingsPage.sidePanelButtonLocator('NES')).click();
    }

    static clickGBSidePanelButton() {
        return findElementBy(SettingsPage.sidePanelButtonLocator('GB')).click();
    }

    static clickGBCSidePanelButton() {
        return findElementBy(SettingsPage.sidePanelButtonLocator('GBC')).click();
    }

    static clickCHIP8SidePanelButton() {
        return findElementBy(SettingsPage.sidePanelButtonLocator('CHIP 8')).click();
    }
}
