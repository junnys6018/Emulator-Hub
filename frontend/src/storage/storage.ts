import { openDB, DBSchema, IDBPDatabase } from 'idb';
import { generateGuestAccount, UserData } from './user-data';
import { v4 as uuidv4 } from 'uuid';

export interface Record {
    uuid: string;
    age: number;
}

export interface EmulatorHubDB extends DBSchema {
    users: {
        value: UserData;
        key: string;
    };
}

export async function createGuestAccount(db: IDBPDatabase<EmulatorHubDB>) {
    console.log('[INFO] No guest account found, generating one');
    const guestUser = await generateGuestAccount();

    const addGuestAccount = async () => {
        try {
            await db.add('users', guestUser);
        } catch (error) {
            if (error.name === 'ConstraintError') {
                console.warn('[WARN] primary key collision, regenerating uuid');
                guestUser.uuid = uuidv4();
                await addGuestAccount();
            } else {
                throw error;
            }
        }
    };
    await addGuestAccount();

    localStorage.setItem('guest-uuid', guestUser.uuid);
    // Dispatch a storage event, this event is not triggered on the window object that modified localStorage
    // so we do it manually
    const event = new StorageEvent('storage', {
        key: 'guest-uuid',
        newValue: guestUser.uuid,
    });
    window.dispatchEvent(event);
}

export async function initializeDatabase() {
    const db = await openDB<EmulatorHubDB>('emulator-hub', 1, {
        upgrade(db, oldVersion) {
            switch (oldVersion) {
                case 0:
                    // database didnt exist, initialize schema
                    console.log('[INFO] Upgrading database to version 1');
                    db.createObjectStore('users', { keyPath: 'uuid' });
            }
        },
        //blocked() {},
        blocking() {
            db.close();
            alert('The application is outdated, please reload the page');
        },
        terminated() {
            alert('[FATAL] database connection was unexpectedly closed');
        },
    });

    // Check if a guest account exists, if not create one
    const guestUuid = localStorage.getItem('guest-uuid');
    if (!guestUuid) {
        await createGuestAccount(db);
    }

    db.close();
}
