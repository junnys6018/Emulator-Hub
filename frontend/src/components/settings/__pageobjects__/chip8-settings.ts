import { By } from 'selenium-webdriver';
import { findElementsBy } from '@/tests/selenium';
import { SettingsPanel } from './settings-page';

export class CHIP8SettingsPanel extends SettingsPanel {
    private static get buttonsLocator() {
        return By.css('div.settings__chip8-grid > *');
    }

    static async keyboardButtons() {
        return findElementsBy(CHIP8SettingsPanel.buttonsLocator);
    }
}
