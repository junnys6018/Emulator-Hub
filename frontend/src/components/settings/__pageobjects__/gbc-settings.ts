import { By } from 'selenium-webdriver';
import { findElementsBy } from '@/tests/selenium';
import { SettingsPanel } from './settings-page';

export class GBCSettingsPanel extends SettingsPanel {
    private static get buttonsLocator() {
        return By.css('div.grid > *');
    }

    static async keyboardButtons() {
        const elements = await findElementsBy(GBCSettingsPanel.buttonsLocator);
        // Select all the keyboard buttons from the grid
        return elements.filter((_element, index) => {
            return index % 3 === 1 && index !== 1;
        });
    }
}
