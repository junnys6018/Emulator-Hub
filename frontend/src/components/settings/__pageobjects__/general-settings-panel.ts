import { By } from 'selenium-webdriver';
import { findElementBy } from '@/tests/selenium';
import { SettingsPanel } from './settings-page';

export class GeneralSettingsPanel extends SettingsPanel {
    private static get showHiddenGamesSwitchLocator() {
        return By.css("input[type='checkbox']");
    }

    static showHiddenGamesSwitch() {
        return findElementBy(GeneralSettingsPanel.showHiddenGamesSwitchLocator);
    }
}
