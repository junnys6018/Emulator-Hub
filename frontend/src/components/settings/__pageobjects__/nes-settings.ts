import { By } from 'selenium-webdriver';
import { findElementsBy } from '@/tests/selenium';
import { SettingsPanel } from './settings-page';

export class NESSettingsPanel extends SettingsPanel {
    private static get buttonsLocator() {
        return By.css('div.settings__controller-grid > *');
    }

    static async keyboardButtons() {
        const elements = await findElementsBy(NESSettingsPanel.buttonsLocator);
        // Select all the keyboard buttons from the grid
        return elements.filter((_element, index) => {
            return index % 3 === 1 && index !== 1;
        });
    }
}
