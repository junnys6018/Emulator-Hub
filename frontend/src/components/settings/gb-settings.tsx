import { defaultGamepadSettings, useUserProfile, GamepadSettings } from '@/src/storage/user-data';
import React, { Fragment } from 'react';
import { SettingsControlGrid, SettingsTitle } from './common';

export default function GBSettings() {
    const [{ settings }, setUserData] = useUserProfile();

    const onChange = (action: string, key: string | number | null) => {
        if (action.endsWith('keyboard')) {
            action = action.replace(/-keyboard$/, '');
            if (key === null) {
                key = defaultGamepadSettings[action as keyof typeof defaultGamepadSettings][0];
            }
            settings.gbSettings[action as keyof GamepadSettings][0] = key as string;
            setUserData({ settings });
        }
    };

    const resetAll = () => {
        // Create a deep clone here
        settings.gbSettings = JSON.parse(JSON.stringify(defaultGamepadSettings));
        setUserData({ settings });
    };

    return (
        <Fragment>
            <SettingsTitle title="Game Boy Controls" />
            <SettingsControlGrid
                controls={
                    settings.gbSettings as unknown as {
                        [action: string]: [string, number];
                    }
                }
                onChange={onChange}
            />
            <button className="btn-secondary mt-auto mb-24 h-10 px-16 w-max" onClick={resetAll}>
                Reset All
            </button>
        </Fragment>
    );
}
