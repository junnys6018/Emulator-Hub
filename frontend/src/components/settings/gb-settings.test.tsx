import _ from 'lodash';
import { driver, resetDriver, elementHasClass } from '@/tests/selenium';
import { SettingsPage } from './__pageobjects__/settings-page';
import { GBSettingsPanel } from './__pageobjects__/gb-settings';
import { By } from 'selenium-webdriver';
import { defaultSettings } from '@/src/storage/user-data';

import React from 'react';
import renderer from 'react-test-renderer';
import GBSettings from './gb-settings';

beforeEach(async () => {
    await driver.navigate().to('http://localhost:8000/settings');
});

afterEach(async () => {
    resetDriver();
});

jest.mock('@/src/storage/user-data', () => {
    const originalModule = jest.requireActual('@/src/storage/user-data');
    const useUserProfile = jest.fn();
    useUserProfile.mockReturnValue([
        {
            profileImage: '/image.png',
            userName: 'Guest',
            settings: {
                gbControls: {
                    up: ['ArrowUp', 12],
                    down: ['ArrowDown', 13],
                    left: ['ArrowLeft', 14],
                    right: ['ArrowRight', 15],
                    a: ['KeyZ', 0],
                    b: ['KeyX', 1],
                    start: ['KeyQ', 9],
                    select: ['Enter', 8],
                },
            },
        },
        jest.fn(),
    ]);
    return {
        __esModule: true,
        ...originalModule,
        useUserProfile,
    };
});

describe('GB Settings', () => {
    test('Click on side panel', async () => {
        await SettingsPage.clickGBSidePanelButton();
        expect(await SettingsPage.getTitle()).toBe('Game Boy Controls');
    });

    test('Initial state', async () => {
        await SettingsPage.clickGBSidePanelButton();

        const keyboardButtons = await GBSettingsPanel.keyboardButtons();
        expect(keyboardButtons.length).toBe(8);
        const names = ['UP ARROW', 'DOWN ARROW', 'LEFT ARROW', 'RIGHT ARROW', 'Z', 'X', 'Q', 'ENTER'];
        for (let i = 0; i < 8; i++) {
            const button = keyboardButtons[i];
            const innerButton = await button.findElement(By.css('button'));
            expect(await innerButton.getText()).toBe(names[i]);
        }

        const saveButton = await GBSettingsPanel.saveButton();
        expect(await elementHasClass(saveButton, 'disabled')).toBe(true);
    });

    test('Changes should be saved', async () => {
        await SettingsPage.clickGBSidePanelButton();

        let keyboardButtons = await GBSettingsPanel.keyboardButtons();
        await keyboardButtons[0].click();

        const body = driver.findElement(By.css('body'));
        await body.sendKeys('=');

        // fetch buttons again
        keyboardButtons = await GBSettingsPanel.keyboardButtons();
        expect(await keyboardButtons[0].getText()).toBe('EQUAL');

        let saveButton = await GBSettingsPanel.saveButton();
        expect(await elementHasClass(saveButton, 'disabled')).toBe(false);

        // Save settings
        await saveButton.click();

        // check indexeddb
        const result = await driver.executeScript(async () => {
            return await new Promise(resolve => {
                const openRequest = indexedDB.open('emulator-hub');
                openRequest.onsuccess = () => {
                    const db = openRequest.result;
                    const transaction = db.transaction('users');
                    const users = transaction.objectStore('users');

                    const request = users.getAll();
                    request.onsuccess = () => {
                        resolve(request.result[0].settings);
                    };
                };
            });
        });

        const expected = _.cloneDeep(defaultSettings);
        expected.gbControls.up[0] = 'Equal';

        expect(result).toEqual(expected);

        // Refresh the page
        await driver.navigate().refresh();
        await SettingsPage.clickGBSidePanelButton();

        keyboardButtons = await GBSettingsPanel.keyboardButtons();
        expect(await keyboardButtons[0].getText()).toBe('EQUAL');

        saveButton = await GBSettingsPanel.saveButton();
        expect(await elementHasClass(saveButton, 'disabled')).toBe(true);
    });

    test('Snapshot', () => {
        const tree = renderer.create(<GBSettings />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
