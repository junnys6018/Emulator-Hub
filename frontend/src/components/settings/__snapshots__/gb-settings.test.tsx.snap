// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GB Settings Snapshot 1`] = `
Array [
  <h2
    className="hidden lg:block text-2xl text-gray-500 mb-7"
  >
    Settings
  </h2>,
  <h1
    className="font-medium text-xl mb-4 lg:text-3xl lg:mb-10"
  >
    Game Boy Controls
  </h1>,
  <div
    className="settings__controller-grid"
  >
    <h3
      className="tracking-wider text-gray-400"
    >
      ACTION
    </h3>
    <h3
      className="tracking-wider text-gray-400 ml-5"
    >
      KEYBOARD
    </h3>
    <h3
      className="tracking-wider text-gray-400 ml-5"
    >
      CONTROLLER
    </h3>
    <span
      className="py-1 my-1.5 tracking-wider"
    >
      UP
    </span>
    <div
      className="settings__control-button  
                "
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        className="tracking-wider text-left flex-grow focus-visible:outline-none "
        onClick={[Function]}
      >
        UP ARROW
      </button>
    </div>
    <div
      className="settings__control-button  
                disabled"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        className="tracking-wider text-left flex-grow focus-visible:outline-none cursor-not-allowed"
        onClick={[Function]}
      >
        DPAD UP
      </button>
    </div>
    <span
      className="py-1 my-1.5 tracking-wider"
    >
      DOWN
    </span>
    <div
      className="settings__control-button  
                "
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        className="tracking-wider text-left flex-grow focus-visible:outline-none "
        onClick={[Function]}
      >
        DOWN ARROW
      </button>
    </div>
    <div
      className="settings__control-button  
                disabled"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        className="tracking-wider text-left flex-grow focus-visible:outline-none cursor-not-allowed"
        onClick={[Function]}
      >
        DPAD DOWN
      </button>
    </div>
    <span
      className="py-1 my-1.5 tracking-wider"
    >
      LEFT
    </span>
    <div
      className="settings__control-button  
                "
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        className="tracking-wider text-left flex-grow focus-visible:outline-none "
        onClick={[Function]}
      >
        LEFT ARROW
      </button>
    </div>
    <div
      className="settings__control-button  
                disabled"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        className="tracking-wider text-left flex-grow focus-visible:outline-none cursor-not-allowed"
        onClick={[Function]}
      >
        DPAD LEFT
      </button>
    </div>
    <span
      className="py-1 my-1.5 tracking-wider"
    >
      RIGHT
    </span>
    <div
      className="settings__control-button  
                "
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        className="tracking-wider text-left flex-grow focus-visible:outline-none "
        onClick={[Function]}
      >
        RIGHT ARROW
      </button>
    </div>
    <div
      className="settings__control-button  
                disabled"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        className="tracking-wider text-left flex-grow focus-visible:outline-none cursor-not-allowed"
        onClick={[Function]}
      >
        DPAD RIGHT
      </button>
    </div>
    <span
      className="py-1 my-1.5 tracking-wider"
    >
      A
    </span>
    <div
      className="settings__control-button  
                "
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        className="tracking-wider text-left flex-grow focus-visible:outline-none "
        onClick={[Function]}
      >
        Z
      </button>
    </div>
    <div
      className="settings__control-button  
                disabled"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        className="tracking-wider text-left flex-grow focus-visible:outline-none cursor-not-allowed"
        onClick={[Function]}
      >
        A
      </button>
    </div>
    <span
      className="py-1 my-1.5 tracking-wider"
    >
      B
    </span>
    <div
      className="settings__control-button  
                "
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        className="tracking-wider text-left flex-grow focus-visible:outline-none "
        onClick={[Function]}
      >
        X
      </button>
    </div>
    <div
      className="settings__control-button  
                disabled"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        className="tracking-wider text-left flex-grow focus-visible:outline-none cursor-not-allowed"
        onClick={[Function]}
      >
        B
      </button>
    </div>
    <span
      className="py-1 my-1.5 tracking-wider"
    >
      START
    </span>
    <div
      className="settings__control-button  
                "
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        className="tracking-wider text-left flex-grow focus-visible:outline-none "
        onClick={[Function]}
      >
        Q
      </button>
    </div>
    <div
      className="settings__control-button  
                disabled"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        className="tracking-wider text-left flex-grow focus-visible:outline-none cursor-not-allowed"
        onClick={[Function]}
      >
        START
      </button>
    </div>
    <span
      className="py-1 my-1.5 tracking-wider"
    >
      SELECT
    </span>
    <div
      className="settings__control-button  
                "
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        className="tracking-wider text-left flex-grow focus-visible:outline-none "
        onClick={[Function]}
      >
        ENTER
      </button>
    </div>
    <div
      className="settings__control-button  
                disabled"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <button
        className="tracking-wider text-left flex-grow focus-visible:outline-none cursor-not-allowed"
        onClick={[Function]}
      >
        SELECT
      </button>
    </div>
  </div>,
  <div
    className="flex items-center mt-auto mb-12 lg:mb-24"
  >
    <button
      className="btn-primary h-10 w-40 xl:w-52 mr-10 lg:mr-20 disabled"
      onClick={[Function]}
    >
      Save
    </button>
    <button
      className="btn-tertiary h-10 w-40 xl:w-52 lg:mr-20 cursor-not-allowed ring-gray-500 text-gray-500"
      onClick={[Function]}
    >
      Cancel
    </button>
    <button
      className="text-gray-300 hover:text-gray-50"
      onClick={[Function]}
    >
      Reset All
    </button>
  </div>,
]
`;
